src/main/java/org/mitre/dsmiley/httpproxy/ProxyServlet.java

```diff
@@ -519,7 +524,35 @@ protected void copyRequestHeader(HttpServletRequest servletRequest, HttpRequest
       } else if (!doPreserveCookies && headerName.equalsIgnoreCase(org.apache.http.cookie.SM.COOKIE)) {
         headerValue = getRealCookie(headerValue);
       }
-      proxyRequest.addHeader(headerName, headerValue);
+      String finalHeaderValue = null;
+      String sessionId = servletRequest.getSession().getId();
+      SoftReference<Set<HttpCookie>> setSoftReference = rootPathCookies.get(sessionId);
```

Comment:
Overall this code lacks some comments/explanations as it's not obvious.  Most especially here where you're creating these soft references.  Why is this being done?  I's it purely for performance (i.e. would things work without it) to cache something expensive?

And I don't like so much code being added to copyRequestHeader; seems like another method would be better.

Reply:
yes. i agree. I have reviewed code and test again.